diff --git a/android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java b/android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java
index b131db5..44f4308 100644
--- a/android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java
+++ b/android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java
@@ -152,7 +152,7 @@ class GetUserMediaImpl {
             }
             WritableMap params = Arguments.createMap();
             params.putString("facing", isFrontFacing ? "front" : "environment");
-            params.putString("deviceId", "" + i);
+            params.putString("deviceId", deviceName);
             params.putString("groupId", "");
             params.putString("label", deviceName);
             params.putString("kind", "videoinput");
diff --git a/src/RTCPeerConnection.ts b/src/RTCPeerConnection.ts
index bb20f59..b69e3ba 100644
--- a/src/RTCPeerConnection.ts
+++ b/src/RTCPeerConnection.ts
@@ -522,6 +522,7 @@ export default class RTCPeerConnection extends defineCustomEventTarget(...PEER_C
 
             if (oldTransceiver) {
                 transceiver = oldTransceiver;
+                transceiver._receiver._track =  new MediaStreamTrack(transceiver._receiver._track);
                 track = transceiver.receiver.track;
                 transceiver._mid = ev.transceiver.mid;
                 transceiver._currentDirection = ev.transceiver.currentDirection;
