diff --git a/android/build.gradle b/android/build.gradle
index 789f6db5a8b6e9a832d9987c0ad8fce935632677..eb310d985e5e450c92e121b854984ed84f9da4a1 100644
--- a/android/build.gradle
+++ b/android/build.gradle
@@ -12,12 +12,11 @@ buildscript {
 apply plugin: 'com.android.library'
 
 android {
-    compileSdkVersion 28
-    buildToolsVersion "23.0.1"
+    compileSdkVersion 32
 
     defaultConfig {
         minSdkVersion 16
-        targetSdkVersion 22
+        targetSdkVersion 32
         versionCode 3
         versionName "2.0.1"
     }
